File: csharp/grade-school/GradeSchool.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class School
{
    // Holds (grade, [student1, student2, ...]) pairs
    private Dictionary<int, List<string>> students = new Dictionary<int, List<string>>();
    
    public void Add(string student, int grade)
    {
        if (this.students.ContainsKey(grade))
        {
            this.students[grade].Add(student);
        }
        else
        {
            this.students.Add(grade, new List<string>() { student });
        }
    }

    public IEnumerable<string> Roster()
    {
        var roster = new List<string>();

        foreach (var grade in this.students.Keys.OrderBy(x => x))
        {
            roster.AddRange(this.students[grade].OrderBy(x => x));
        }

        return roster;
    }

    public IEnumerable<string> Grade(int grade)
    {
        if (this.students.ContainsKey(grade))
        {
            return this.students[grade].OrderBy(x => x);
        }
        else
        {
            return new List<string>();
        }
    }
}

File: ruby/grade-school/grade_school.rb
class School  
  def initialize
    @grade_levels = Hash.new { |h, k| h[k] = [] }
  end
  
  def add(student, grade)
    @grade_levels[grade].push(student).sort!
  end

  def students(grade)
    @grade_levels[grade]
  end

  def students_by_grade
    @grade_levels.sort.map { |k, v| { grade: k, students: v } }.dup
  end
end
